enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,87 @@
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM misfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect in-active cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
+to: `Operator checking for stale status, the active operator will reset this message: <original message>`.
+
+After the message is changed, as it suggests, the active operators should reset this message back to the original
+message, removing the "checking" part. This shows a healthy, active operator.
+
+Suppose the operator is not active and does not reset the message back to the original after 10 minutes. In that case,
+the controller will change the message again
+to: `Operator has not updated this condition for more than 20 minutes, last known condition state was "<False/True>", original message: <original message>`
+. In addition to changing the message, this time, the controller will also change the condition status from False or
+True to **Unknown**.
+
+For cluster administrators, this means the operator is having a problem and is potentially down or unable to make
+updates to conditions. In that case, an alert might be required to trigger a manual investigation of the current
+operator state.
+
+### API Extensions
+
+*No API extensions required*
+
+### Risks and Mitigations
+
+One risk here is that we find operators that are not correctly reconciling their clusteroperator conditions, which this
+controller makes obvious. These operators should be fixed as a result.
+
+## Alternatives
+
+One alternative discussed was a new API field that will be added to clusteroperators.Status called "lastUpdateTime".
+Instead of relying on
+"lastTransitionTime" will introduce this field and use it as "heart-beat". This will require **all** operators to update
```

Comment:
this approach would also multiply writes significantly.

Reply:
more significantly than the proposed mechanism?  in either case isn't a healthy operator going to be doing a write to clusteroperator every 10 minutes?  (either to update the heartbeat, or to re-assert "no i'm not stale")